data/virt_autotest/guest_console_monitor.sh

```diff
@@ -0,0 +1,236 @@
+}
+
+#Quit if there are less than one argument
+if [ $# -eq 0 ];then
+    help_usage
+    exit 1
+fi
+
+#PREPARATION
+##CONFIG PART
+START_MONITOR="0"
+END_MONITOR="0"
+LOG_DIR="/tmp/virt_logs_residence"
+MONITOR_LOCKED_FILE="/tmp/GUEST_CONSOLE_COLLECTOR_SEMAPHORE"
+timestamp="`date '+%Y%m%d%H%M%S'`"
+###CHECK EXISTED GUEST
+vm_guestnames_types="sles"
+get_vm_guestnames_inactive=`virsh list --inactive | grep "${vm_guestnames_types}" | awk '{print $2}'`
+vm_guestnames_inactive_array=$(echo -e ${get_vm_guestnames_inactive})
+get_vm_guestnames=`virsh list  --all | grep "${vm_guestnames_types}" | awk '{print $2}'`
+vm_guestnames_array=$(echo -e ${get_vm_guestnames})
+vmguest=""
+vmguest_failed="0"
+
+#Parse input arguments. -s or -e must have values
+while getopts "seh" OPTION; do
+  case "${OPTION}" in
+    s)
+      START_MONITOR="1"
+      ;;
+    e)
+      END_MONITOR="1"
+      ;;
+    h)
+      help_usage
+      exit 1
+      ;;
+    *)
+      help_usage
+      exit 1
+      ;;
+  esac
+done
+
+#FUNCTION PART
+function get_console() {
+
+local vmguest=$1
+
+expect -c "
+set timeout 60
+
+##hide echo
+log_user 0
+spawn -noecho virsh console ${vmguest}
+
+#wait connection
+sleep 3
+send \"\r\n\r\n\r\n\"
+
+#condition expect
+expect {
+        \"*login:\" {
+                send \"root\r\"
+                exp_continue
+        }
+        -nocase "password:" {
+                send \"novell\r\"
+                exp_continue
+        }
+        \"*:~ #\" {
+                send -- \"ip route get 1\r\"
+        }
+        timeout {
+                send -- \"exit\r\"
+                exp_continue        
+        }                 
+}
+
+## -1 means never timeout
+set timeout -1
+
+expect -re {dev.*\s([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})[^0-9]}
+
+if {![info exists expect_out(1,string)]} {
+        puts \"Match did not happen :(\"
+        exit 1
+}
+set output \$expect_out(buffer)
+expect eof
+
+puts \"\$output\"
+"
+
+}
+
+function open_console() {
+
+    #open guest console
+    for vmguest in ${vm_guestnames_state_running_array[@]};do
+        get_console ${vmguest} >> ${LOG_DIR}/${vmguest}_console_${timestamp} 2>&1 &
+    done 
+}
+
+function close_console() {
+
+    #close guest console
+    echo "kill PIDs $(basename $0)"
+    pid=`ps aux | grep "$(basename $0) -s" | grep -v "grep"| awk '{print $2}'`
+    if [[ -n $pid ]];then
+        kill -9 $pid >> /dev/null 2>&1;sync
+    else
+        echo "PIDs $(basename $0) were killed"
+    fi
+
+    echo "Kill all associated "virsh console" processes"
+    ps aux | grep "virsh console" | grep -v "grep" >> /dev/null 2>&1
+    while [[ $? -eq 0 ]];
+    do
+        #wait for checking all associated "virsh console" processes"
+        sleep 3
+        sub_pid=`ps aux | grep "expect -c" | grep -v "grep"| awk '{print $2}'`
+        if [[ -n $sub_pid ]];then
+            kill -9 $sub_pid >> /dev/null 2>&1;sync
```

Comment:
Failing to kill will also terminate the loop ?

Reply:
Refer to the output from # pstree -a:
```
# pstree -a
...
  ├─guest_console_m ./guest_console_monitor.sh -s
  │   ├─guest_console_m ./guest_console_monitor.sh -s
  │   │   └─expect -c...
  │   │       ├─virsh console sles-15-sp2-64-fv-def-net
  │   │       │   └─{virsh}
  │   │       └─{expect}
  │   ├─guest_console_m ./guest_console_monitor.sh -s
  │   │   └─expect -c...
  │   │       ├─virsh console sles-15-sp2-64-pv-def-net
  │   │       │   └─{virsh}
  │   │       └─{expect}
```

guest_console_monitor.sh -s(parent process) -> expect -c (child process) ->  "virsh console"

Reply:
Yes. But failure of "kill -9 $sub_pid >> /dev/null 2>&1;sync" with "-n $sub_pid" also jump out of while loop. So the final result is  "-n $sub_pid" and $sub_pid was not killed.

Reply:
Great, will update this.

Reply:
Are you going to update this ?

Reply:
Quick updated, after `kill -9 $sub_pid >> /dev/null 2>&1;sync`, which was still ran in while loop without  jump out of while loop.

Reply:
DONE